### PR TITLE
fix: add transaction support for dgraph migration

### LINKED ISSUE #2892: Transaction in dGraph Migration
A lot of datasources in GoFr do not have transaction like logic implemented. This needs to be resolved for better consistency. This ticket is one of the many that addresses one of the various migration supported databases.

### PR DESCRIPTION
**Pull request by krisp619**
Fixes #2892

Summary:
This change adds transaction support to Dgraph migration commit flow to ensure consistency during migration execution.

Changes:
- Wrapped mutation logic inside Dgraph transaction
- Added txn interface with Mutate / Commit / Discard
- Used NewTxn + defer Discard pattern
- Commit on success, discard on error 
- Updated migration tests

### CODE REVIEW

**Review by coolwednesday — COMMENTED** [review]
Thanks for picking this up, @krisp619. The transaction wrapping pattern is correct and the test structure is solid. However, there are a few issues that need to be addressed before merge.

---

### Blocking: Loss of observability on mutation

**File:** `pkg/gofr/migration/dgraph.go:184`

The original code called `c.DGraph.Mutate()` which goes through `Client.Mutate()` ([dgraph.go:205-237](https://github.com/gofr-dev/gofr/blob/development/pkg/gofr/datasource/dgraph/dgraph.go#L205-L237)). That method:
- Creates an **OpenTelemetry span** via `addTrace(ctx, "mutate")`
- **Logs** mutation failures via `d.logger.Error`
- **Records histogram metrics** (`dgraph_mutate_duration`) via `sendOperationStats`
- **Pretty-prints** the query log

The new `tx.Mutate()` calls `txnImpl.Mutate()` directly — a raw pass-through to the dgo client with **no tracing, no logging, no metrics**. Migration mutations become invisible to the observability stack.

**Options:**
- (a) Add tracing/logging/metrics to the transaction-based mutation path to match existing behaviour
- (b) At minimum, log mutation duration and errors at the migration layer
- (c) Document this as a known trade-off if observability on migration mutations is intentionally deprioritized

---

###  `Discard` error silently dropped

**File:** `pkg/gofr/migration/dgraph.go:182`

```go
defer tx.Discard(ctx)
```

`Discard` returns an `error` that is silently dropped. While this matches the idiomatic `dgo` pattern, the framework logs errors from other datastores. Consider:

```go
defer func() {
    if dErr := tx.Discard(ctx); dErr != nil {
        c.Errorf("dgraph: transaction discard failed: %v", dErr)
    }
}()
```

---

### Missing test coverage for edge cases

**File:** `pkg/gofr/migration/dgraph_test.go`

The following edge cases are not covered:

| # | Edge Case | Why It Matters |
|---|---|---|
| 1 | `NewTxn()` returns a type that doesn't implement `dgraphTxn` | `errInvalidDgraphTxn` path is never exercised. Test with: `mockDGraph.EXPECT().NewTxn().Return("not-a-txn")` |
| 2 | `NewTxn()` returns `nil` | nil type assertion should yield `ok=false`, verify it returns `errInvalidDgraphTxn` and doesn't panic. Test with: `mockDGraph.EXPECT().NewTxn().Return(nil)` |
| 3 | `Discard` returns an error | Verify no panic and the commit flow still returns the correct error (mutate or commit error, not discard error) |
| 4 | `Discard` called after successful `Commit` | Verify dgo no-op behaviour — `Discard` after `Commit` should not corrupt data |
| 5 | `json.Marshal` fails before `NewTxn()` | Verify no transaction is leaked (txn never created). Pre-existing path but good to confirm. |

---

### Minor style nits

1. **`var ( errInvalidDgraphTxn = ... )`** — Single var doesn't need parentheses. Use: `var errInvalidDgraphTxn = errors.New("...")`

2. **`commitDone` field in `mockDgraphTxn`** — This is set to `true` even when commit returns an error. The name implies success. Consider renaming to `commitCalled` for clarity.

3. **Error message** — `"invalid dgraph transaction type"` gives no debugging info. Consider:
   ```go
   return fmt.Errorf("dgraph: NewTxn() returned %T, expected dgraphTxn", c.DGraph.NewTxn())
   ```

---

### Please add screenshots/logs showing

The PR description should include evidence of correct transaction behaviour. Specifically:

**1. Successful migration run**
- Start a GoFr application with Dgraph configured and a pending migration
- Show logs confirming: transaction created → mutate succeeded → commit succeeded
- Query Dgraph to verify the migration record exists:
  ```graphql
  {
    migrations(func: type(Migration), orderdesc: migrations.version, first: 1) {
      migrations.version
      migrations.method
      migrations.start_time
      migrations.duration
    }
  }
  ```
- Screenshot of the query result showing the migration was recorded

**2. Failed mutation (rollback verification)**
- Simulate a mutation failure (e.g., malformed JSON payload, network error, or Dgraph down)
- Show logs confirming: transaction created → mutate failed → discard called → migration rolled back
- Query Dgraph to verify **no partial migration record** was committed

**3. Failed commit (post-mutate failure)**
- Simulate a commit failure (e.g., transaction conflict / network error after mutate)
- Show logs confirming: transaction created → mutate succeeded → commit failed → discard called
- Query Dgraph to verify the mutation was **not persisted** (transaction was discarded)

**4. Idempotency check**
- Run the same migration twice
- Show that the second run detects the migration is already applied (via `getLastMigration` query) and skips it
- This confirms the commit path writes the correct version number

**5. Before/after comparison (optional but helpful)**
- Run a migration on the `development` branch (without this PR) and show Dgraph server logs
- Run the same migration on this branch and show the difference — explicit `BeginTxn` → `Commit` sequence visible in logs

---

### Edge cases to test against (manual or integration)

Beyond the unit tests listed above, these should be verified against a real Dgraph instance:

| # | Scenario | What to check |
|---|---|---|
| 1 | Dgraph server goes down between `Mutate` and `Commit` | Error is returned, no partial data, app doesn't panic |
| 2 | Context cancelled during `Mutate` | Transaction is discarded, migration is marked as failed |
| 3 | Large migration payload (many predicates/fields) | Mutation + commit complete within timeout, no OOM |
| 4 | Multiple GoFr instances running migrations concurrently | Transaction conflicts are handled gracefully (dgo returns `ErrAborted`) |
| 5 | Dgraph returns `ErrAborted` (transaction conflict) | Error is propagated, migration can be retried |
| 6 | Schema migration (`ApplySchema`) succeeds but metadata commit fails | Schema change is applied but migration is not recorded — next run will retry. Document this as expected behaviour. |

---

### Summary

The core transaction logic is correct — `NewTxn → Mutate → Commit` with `defer Discard` is the right pattern. The blocking issue is the **loss of tracing/logging/metrics** on the mutation path. Once that's addressed and the missing test cases are added, this is good to merge.

**Comment by Umang01-hash** [comment]
Hey @krisp619 !
In order to move forward with this PR please resolve the review comments given by @coolwednesday. If you need any help feel free to ask here!

**Comment by krisp619** [comment]
@Umang01-hash @coolwednesday Thanks for the detailed review.

I’ve been going through the suggested changes and verifying the current implementation,
but I still need some time to properly address the observability and additional test cases mentioned.

I’ll continue working on this and push the updates soon.

**Comment by Umang01-hash** [comment]
Hey @krisp619 Your PR contains some linter error, please resolve it so we can go ahead and review the PR.

Screenshot 2026-03-27 at 5 12 15 PM

**Comment by krisp619** [comment]
@coolwednesday

Updated.

* Transaction flow (NewTxn → Mutate → Commit → Discard) implemented
* Safe discard with error logging added
* Observability restored via logging
* Added test cases for:

  * invalid transaction type
  * mutation failure
  * commit failure

Currently resolving a small test syntax issue, after which I’ll re-run tests and update.

Please let me know if anything else is needed.

**Comment by krisp619** [comment]
@coolwednesday

All requested changes have been implemented.

* Transaction flow fixed (NewTxn → Mutate → Commit → Discard)
* Safe discard with logging added
* Observability restored
* Added test cases for failure scenarios

Dgraph-specific tests are passing locally.

Please let me know if anything else is needed.
Screenshot 2026-03-31 004529

**Comment by krisp619** [comment]
@coolwednesday

Just a gentle follow-up. The merge conflicts have been resolved and the branch is updated.

Please let me know if any further changes are needed from my side.

**Comment by krisp619** [comment]
@coolwednesday @Umang01-hash 

Hi, just following up on this PR.

All requested changes have been implemented, conflicts are resolved, and the branch is up to date.

Could you please review when you get a chance? Happy to make any further updates if needed.

Thanks!

**Comment by Umang01-hash** [comment]
@krisp619 sure thankyou will review your pr soon.

**Comment by Umang01-hash** [comment]
Hey @krisp619, thanks for the transaction work here 🙏 I pushed onto your branch to help get it across the line — your commits are untouched, mine are added as a co-author.

While verifying end-to-end against a real Dgraph, the transaction alone still re-ran every migration on restart. There were two more issues in the existing dgraph migrator:

- the migration record wasn't tagged with `dgraph.type: "Migration"`, so the `type(Migration)` lookup in `getLastMigration` never found it
- `getLastMigration` decoded the wrong shape — the result is in `api.Response.Json` and the predicate is `migrations.version`, so it always returned 0

Added those next to your transaction commit/discard, and kept the `UsedDatasources` guard so we only record when a migration actually touches Dgraph. Also merged latest `development` in so the branch is mergeable again.

Verified locally with a docker Dgraph: first run applies the migrations, second run logs `no new migrations to run`, and no duplicate records are written. Unit tests + lint + vet all green.

Closes #2892.

File: pkg/gofr/migration/dgraph.go [hunk 1]
```diff
@@ -11,6 +12,19 @@ import (
...
+// migrationMethodUP is the method recorded for an applied (UP) migration.
+const migrationMethodUP = "UP"
+
+// dgraphTxn is the subset of the Dgraph transaction used to record migrations.
+// The value returned by Dgraph.NewTxn satisfies it at runtime.
+type dgraphTxn interface {
+	Mutate(ctx context.Context, mu *api.Mutation) (*api.Response, error)
+	Commit(ctx context.Context) error
+	Discard(ctx context.Context) error
+}
```
[the remark this answers is not in the document]
Comment (Umang01-hash) [reply pkg/gofr/migration/dgraph.go:26]: Verified against the code — this isn't a real issue. `c.DGraph.NewTxn()` returns the datasource's `*txnImpl` wrapper (pkg/gofr/datasource/dgraph/interfaces.go), not the raw `*dgo.Txn`, and `txnImpl.Mutate` is `(ctx, *api.Mutation) (*api.Response, error)` — matching `dgraphTxn` exactly, so the assertion succeeds. (dgo's own `Txn.Mutate` also returns `*api.Response`, not `*api.Assigned`.) Confirmed end-to-end on a real Dgraph: records commit and a second run logs "no new migrations to run", which only happens if the assertion succeeds.

File: pkg/gofr/migration/dgraph_test.go [hunk 2]
```diff
@@ -13,6 +14,33 @@ import (
...
+	committed bool
+	discarded bool
+}
+
+func (f *fakeDgraphTxn) Mutate(_ context.Context, mu *api.Mutation) (*api.Response, error) {
+	f.mutated = true
+	f.setJSON = mu.SetJson
+
+	return nil, f.mutateErr
+}
```
[the remark this answers is not in the document]
Comment (Umang01-hash) [reply pkg/gofr/migration/dgraph_test.go:32]: The fake matches production, not the raw dgo txn: `c.DGraph.NewTxn()` yields `*txnImpl`, whose `Mutate` returns `(*api.Response, error)` — same as the fake. dgo's txn `Mutate` also returns `*api.Response` (not `*api.Assigned`), so there's no production divergence to mask.

File: pkg/gofr/migration/dgraph.go [hunk 3]
```diff
@@ -161,13 +174,25 @@ func (dm dgraphMigrator) commitMigration(c *container.Container, data transactio
 		return err
 	}
 
-	_, err = c.DGraph.Mutate(context.Background(), &api.Mutation{
+	tx, ok := c.DGraph.NewTxn().(dgraphTxn)
+	if !ok {
+		return errInvalidDgraphTxn
+	}
```
[the remark this answers is not in the document]
Comment (Umang01-hash) [reply pkg/gofr/migration/dgraph.go:187]: Already covered — `Test_DGraphCommitMigration` has an "invalid transaction type" case that makes `NewTxn()` return a non-transaction value and asserts `errInvalidDgraphTxn` is returned.

**Review by aryanmehrotra — COMMENTED** [review]
Reviewed at `423761c6a`. The change is correct and it fixes a real, severe bug — but the bug is bigger than the PR title says, and the write-up undersells what is actually being repaired here.

**Gates:** `go build ./...` clean · `go test ./pkg/gofr/migration/...` pass · `gofmt` clean · `go vet` clean · `golangci-lint --new-from-rev=origin/development` **0 issues**.

## What is actually broken on `development`

I went looking for whether wrapping a single mutation in an explicit transaction buys anything — a lone Dgraph mutation is already atomic, so on the face of it "add transaction support" reads like a no-op. It isn't. There are **three independent reasons** a Dgraph migration record never survives today, and this PR fixes all three:

**1. The mutation is never committed.** `pkg/gofr/datasource/dgraph/dgraph.go:220`:

```go
resp, err := d.client.NewTxn().Mutate(tracedCtx, mutation)
```

No `Commit`, no `Discard`, and the caller's `&api.Mutation{SetJson: ...}` leaves `CommitNow` false. Verified against the pinned dgo (`v210.0.0-20230328113526-b66f8ae53a2d`, `txn.go:153`): `Mutate` forwards `CommitNow: mu.CommitNow` and `txn.go:206` only finishes the transaction when it is set. So the mutation is staged and the transaction is abandoned to the GC. **Nothing is persisted.**

**2. The record has no `dgraph.type`.** `getLastMigration` queries `func: type(Migration)`, so even a committed record would never be found.

**3. The query and the struct disagree.** The response struct expects `json:"version"`, the query selected `migrations.version` with no alias, and `getLastMigration` then did `json.Marshal(resp)` on an `any` holding `*api.Response` before unmarshalling — which yields the wrapper, not the payload.

Any one of those makes `getLastMigration` return 0 forever, so **every Dgraph migration re-runs on every boot**. The PR fixes all three: the txn + `Commit`, `"dgraph.type": "Migration"`, and the `version:` alias with direct `r.Json` decoding.

## The main thing I'd like addressed

🔴 **Point 1 is a data-loss bug in a public API, and this PR routes around it instead of fixing it.**

`Client.Mutate` is what application code calls as `ctx.DGraph.Mutate(...)`. It has never committed. This PR stops the *migration* using it — by calling `NewTxn()` directly — and leaves every other caller silently dropping writes. Inside `pkg/` the migration was the only caller, so the diff looks complete; outside it, every GoFr user mutating Dgraph is affected.

Two ways to close it, and I'd take either:

- fix `Client.Mutate` to commit (set `CommitNow`, or `NewTxn()` + `Commit`), which also makes the migration change a genuine consistency improvement rather than a workaround; or
- file it and link the issue here, so it doesn't get buried by a green PR that appears to have handled Dgraph mutations.

What I'd rather not see is this merging with the public path still broken and no record of it.

## Non-blocking

**The type assertion has no compile-time guarantee and nothing exercises it.**

```go
tx, ok := c.DGraph.NewTxn().(dgraphTxn)
```

`container.Dgraph.NewTxn()` returns `any`, and the concrete value is `*txnImpl` from **`pkg/gofr/datasource/dgraph`, a separate Go module**. I checked it holds today — `txnImpl.Mutate/Commit/Discard` match `dgraphTxn` exactly, and both modules pin the identical dgo version — so this is not a live bug.

But it holds by two independently-versioned modules happening to agree. The datasource module can bump dgo on its own; if `*api.Mutation` changes shape, the assertion starts failing at runtime and **every Dgraph migration errors with `errInvalidDgraphTxn`**, with nothing in CI to catch it: `fakeDgraphTxn` is defined in the root module's test and satisfies the interface by construction, so it proves the logic and says nothing about the real type. The previous code had no such coupling — `Mutate(ctx, mu any) (any, error)` is `any`-typed at the boundary.

Since a cross-module compile-time assertion isn't available, at minimum make the failure diagnosable. `"invalid Dgraph transaction"` gives an operator nothing:

```go
errInvalidDgraphTxn = errors.New(
    "dgraph migration: NewTxn() did not return a transaction implementing " +
    "Mutate/Commit/Discard — check that the dgraph datasource module's dgo version matches")
```

Better still would be declaring `dgraphTxn` somewhere both modules can see it (next to `container.Dgraph`) so the datasource can assert against it at compile time.

**The description undersells the change.** "Adds transaction support to Dgraph migration commit flow" describes the smallest of the three fixes. `dgraph.type` and the query alias are what actually make `getLastMigration` work, and the missing commit is a data-loss fix. Someone reading the changelog to decide whether to upgrade cannot tell that "Dgraph migrations re-run from zero on every boot" was the symptom. Worth restating the body around that.

**Two smaller ones:**

- `ctx := context.Background()` for the mutation and commit — pre-existing in this function, but now that the operation spans mutate-then-commit it is more exposed. A migration that hangs here has no deadline and no cancellation.
- `defer tx.Discard(ctx)` after a successful `Commit` is safe — dgo `txn.go:288 commitOrAbort` early-returns `nil` when `txn.finished`, which `Commit` has already set — so the deferred handler will not log a spurious error. I checked because the `c.Errorf` in the defer would otherwise fire on every successful migration. **No change needed**; noting it so the next reader doesn't have to re-derive it.

## Tests

The rewrite from the table over `map[string]any` to a `fakeDgraphTxn` with explicit success / mutate-error / commit-error cases is a real improvement, and asserting the payload carries `dgraph.type` and the version is the right assertion — it pins the thing that was broken, not just the happy path.

The gap is the one above: nothing tests the assertion against the real `*txnImpl`. That is inherent to the module split rather than something to fix in this PR.

---

Nice find on the root cause. The transaction wrapping is correct, and the two undeclared fixes are the ones that actually make Dgraph migrations work — they deserve to be in the title.

**Review by aryanmehrotra — CHANGES_REQUESTED** [review]
I ran this end to end against a real Dgraph v21.03.0 rather than reading the diff, and the change is correct and considerably more valuable than the title suggests. Two things to fix before it goes in, both small.

**This fixes three bugs, not one, and together they mean Dgraph migrations do not work at all on `development`.**

1. `Client.Mutate` runs `d.client.NewTxn().Mutate(tracedCtx, mutation)` and never commits or discards, and the migration payload sets no `CommitNow` — so the record is staged into a transaction that is then abandoned.
2. The payload carries no `dgraph.type`, so even a committed record would be invisible to `getLastMigration`'s `func: type(Migration)`.
3. `getLastMigration` marshalled the `*api.Response` wrapper rather than reading its `Json` field, and the query did not alias the predicate to `version`.

The observable result is that every Dgraph migration re-runs on every application start, forever, and nothing is ever recorded — while the logs report `Inserted record for migration 1 in Dgraph migrations` and `Migration 1 ran successfully`. That is worth calling out in the PR description, because "add transaction support" undersells it.

Three consecutive boots of the same app against a real Dgraph, with a migration that touches DGraph but does not redefine the migration schema:

| | `development` | this PR |
|---|---|---|
| migration body ran, runs 1 / 2 / 3 | yes / yes / yes | yes / no / no |
| records in Dgraph after 3 runs | 0 | 1 |
| record contents | — | `version:1, method:"UP", start_time:2026-09-07T08:23:21Z, duration:5` |

`start_time` decodes correctly against the declared `datetime` predicate, and the version is read back on the next boot, so the idempotency path genuinely closes.

| check | result |
|---|---|
| `pkg/gofr/migration` suite | pass |
| coverage | 89.4% → 89.5% |
| `gofmt` / `go vet` / lint on changed files | clean |
| exported API | unchanged — every new identifier is unexported |

---

**1. The lost tracing, logging and metrics need addressing.** This was already raised and it holds up when measured: counting datasource operation log lines from run 1 of the same app, `development` emits `2 Alter, 1 mutation, 2 query` and this branch emits `2 Alter, 0 mutation, 2 query`. `Client.Mutate` wraps the call in `addTrace`, a `QueryLog` with duration, and a `dgraph_mutate_duration` histogram; `NewTxn()` returns the bare dgo transaction with none of that, so the migration write becomes invisible to the observability stack.

Worth being fair about what is being traded: the instrumented path was reporting a write that never actually landed, so this is a correct-but-silent write replacing an observable-but-fictional one. That is still the right direction, but the migration record write should be traced. `dgraph_mutate_duration` is not referenced by any dashboard in the repo, so there is no dashboard work attached — logging the duration and any error at the migration layer would be enough.

**2. The `NewTxn()` nil guard only covers half the case.** The type assertion behaves differently for the two kinds of nil:

```
untyped nil            -> assertion ok=false -> errInvalidDgraphTxn   (safe)
typed nil (*dgo.Txn)   -> assertion ok=true  -> Discard panics: invalid memory address or nil pointer dereference
```

A typed nil satisfies the interface, so the guard passes and the deferred `Discard` dereferences it. Only reachable from a custom `container.Dgraph` whose `NewTxn()` returns a typed nil pointer, so it is not a likely path, but it is a one-line check.

---

**One thing to flag for the release notes rather than to change here.** There is no exported API change — `container.Dgraph` already declared `NewTxn() any`, `migration.DGraph` is untouched, and every identifier this PR adds is unexported. But the migration path now *calls* `NewTxn()`, which it never did before, and requires the result to implement `Mutate`/`Commit`/`Discard`. Anyone running a hand-written `container.Dgraph` — a test double, or a wrapper around the client — will now see migrations fail with `errInvalidDgraphTxn` where they previously appeared to succeed. That is a runtime break rather than a compile-time one, which is the harder kind to notice, so it deserves a line in the changelog.

The behaviour change for ordinary users is benign and is the point of the fix: migrations that have been quietly re-running on every boot will run once more, get recorded, and then stop.

Core logic is right — `NewTxn` → `Mutate` → `Commit` with a deferred `Discard`, the `dgraph.type` addition and the response decoding are all correct. Happy to approve once the two items above are in.

**Comment by Umang01-hash** [comment]
Thanks for the thorough end-to-end review @aryanmehrotra — both items are in (pushed on top):

**1. Observability.** The txn write bypasses `Client.Mutate`, so the migration layer now logs the record-write duration (`recorded migration <n> in Dgraph in <dur>`, DEBUG — same level as the datasource query logs) and logs mutate/commit errors with migration context before returning. Verified against a real Dgraph: the duration line now shows on the run that records the migration.

**2. Typed-nil guard.** `NewTxn().(dgraphTxn)` returns `ok=true` for a typed nil, so the deferred `Discard` would deref it. Now guarded with the package's `isNil` helper (`if !ok || isNil(tx)`), and added a `typed-nil transaction does not panic` test asserting `errInvalidDgraphTxn`.

**3. Release note** — flagging as requested (runtime, not compile-time break):

> Dgraph migrations are now recorded correctly and no longer re-run on every startup. The migration path now calls `Dgraph.NewTxn()` and requires the returned value to implement `Mutate`/`Commit`/`Discard`. A hand-written `container.Dgraph` (e.g. a test double or client wrapper) whose `NewTxn()` does not return such a transaction will now fail migrations with `invalid Dgraph transaction` instead of silently appearing to succeed.

`pkg/gofr/migration` suite + `gofmt`/`go vet`/lint clean on the changed files.

**Review by aryanmehrotra — APPROVED** [review]
Re-reviewed at `7d8438f0e`. **Both items from the previous round are in, and I verified each rather than reading the diff.** Approving.

### 1. Observability — done

The record write now logs its duration at the migration layer, confirmed firing on a real Dgraph:

```
recorded migration 1 in Dgraph in 25.690875ms
```

plus `c.Errorf` with the migration number on both the mutate and the commit failure path.

Stating the trade explicitly so it is on the record rather than made silently: the structured `mutation … DGRAPH` QueryLog line and the `dgraph_mutate_duration` histogram are still bypassed, because the txn path does not go through `Client.Mutate`. Datasource op lines on run 1 go from `2 Alter, 1 mutation, 2 query` on `development` to `2 Alter, 0 mutation, 2 query` here. That is the trade I asked for — the instrumented line was reporting a write that never landed, and duration plus errors at the migration layer is the right replacement. No dashboard in the repo references that histogram.

### 2. Typed-nil guard — done, and the test earns its place

`if !ok || isNil(tx)`, reusing the existing `isNil` at `pkg/gofr/migration/migration.go:381` rather than adding a helper. Good call.

I checked the new test actually fails without the guard — deleted `|| isNil(tx)` and re-ran just that subtest:

```
--- FAIL: Test_DGraphCommitMigration/typed-nil_transaction_does_not_panic
panic: runtime error: invalid memory address or nil pointer dereference
```

### End-to-end, re-run at this head

Three boots of the same app against Dgraph v21.03.0, `drop_all` between the two sets:

| | `development` | this PR |
|---|---|---|
| migration body ran, runs 1 / 2 / 3 | yes / yes / yes | yes / no / no |
| records after 3 runs | 0 (`"q":[]`) | 1 |
| record | — | `version:1, method:"UP", start_time:2026-09-07T10:10:20Z, duration:30` |
| ERROR lines | — | 0 |

| gate | result |
|---|---|
| `gofmt -l` | clean |
| `go build ./...` | clean |
| `go vet` | clean |
| `go test ./pkg/gofr/migration/...` | pass, 6/6 dgraph subtests |
| coverage | 89.5% |
| `golangci-lint --new-from-rev=origin/development` | 0 issues |

### The one thing I raised that this PR correctly does not fix

`Client.Mutate` never commits, so a caller who omits `CommitNow` gets a silent no-op — which is what the migration code was doing and one of the three reasons migrations never persisted. This PR routes the migration around it, which is right; fixing the public method belongs in its own change.

Filed as #4157, with the pinned-dgo reasoning and a reproduction showing `PERSISTED: without CommitNow = 0, with CommitNow = 1`. So it is on the record rather than buried by this PR going green.

### Follow-ups, none blocking

- **The title still undersells it.** `fix: add transaction support for dgraph migration` describes the smallest of the three fixes. The symptom was that Dgraph migrations re-ran from zero on every boot and never recorded anything, while logging success. Worth putting that in the description for whoever reads the changelog.
- `context.Background()` across mutate→commit has no deadline or cancellation.
- `errInvalidDgraphTxn`'s message gives an operator nothing to act on.
- Runtime-break note for the release notes: migrations now call `NewTxn()` and require the result to implement `Mutate`/`Commit`/`Discard`, so anyone running a hand-written `container.Dgraph` will see `errInvalidDgraphTxn` where migrations previously appeared to succeed.

Good fix, and the root-cause work behind it was the valuable part.
